iroh/src/watcher.rs

```diff
@@ -304,13 +322,133 @@ impl<S: Watcher, T: Watcher> Watcher for (S, T) {
     }
 }
 
+/// Combinator to join two watchers
+#[derive(Debug, Clone)]
+pub struct Join<T: Clone + Eq, W: Watcher<Value = T>> {
+    watchers: Vec<W>,
+}
+impl<T: Clone + Eq, W: Watcher<Value = T>> Join<T, W> {
+    /// Joins a set of watchers into a single watcher
+    pub fn new(watchers: impl Iterator<Item = W>) -> Self {
+        let watchers: Vec<W> = watchers.into_iter().collect();
+
+        Self { watchers }
+    }
+}
+
+impl<T: Clone + Eq + std::fmt::Debug, W: Watcher<Value = T>> Watcher for Join<T, W> {
+    type Value = Vec<T>;
+
+    fn get(&self) -> Result<Self::Value, Disconnected> {
+        let mut out = Vec::with_capacity(self.watchers.len());
+        for watcher in &self.watchers {
+            out.push(watcher.get()?);
+        }
+
+        Ok(out)
+    }
+
+    fn poll_updated(
+        &mut self,
+        cx: &mut task::Context<'_>,
+    ) -> Poll<Result<Self::Value, Disconnected>> {
+        let mut new_value = None;
+        for (i, watcher) in self.watchers.iter_mut().enumerate() {
+            match watcher.poll_updated(cx) {
+                Poll::Pending => {}
+                Poll::Ready(Ok(value)) => {
+                    new_value.replace((i, value));
+                    break;
+                }
+                Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
+            }
+        }
+
+        if let Some((j, new_value)) = new_value {
+            let mut new = Vec::with_capacity(self.watchers.len());
+            for (i, watcher) in self.watchers.iter().enumerate() {
+                if i != j {
+                    new.push(watcher.get()?);
+                } else {
+                    new.push(new_value.clone());
+                }
+            }
+            Poll::Ready(Ok(new))
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+/// Combinator to join two watchers
+#[derive(Debug, Clone)]
+pub struct JoinOpt<T: Clone + Eq, W: Watcher<Value = Option<T>>> {
+    watchers: Vec<W>,
+}
```

Comment:
Continuation from another comment: I'd be happier if we didn't have this, and instead advise using `.map(|vec| vec.into_iter().flatten().collect())`

Reply:
(`.map()` isn't all you need, you also need `Join::new` in advance. I'm totally in favor of keeping `Join`!)